x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/privilege/RestGetBuiltinPrivilegesAction.java

```diff
@@ -73,6 +73,10 @@ public RestResponse buildResponse(GetBuiltinPrivilegesResponse response, XConten
                     builder.startObject();
                     builder.array("cluster", translatedResponse.getClusterPrivileges());
                     builder.array("index", translatedResponse.getIndexPrivileges());
+                    String[] remoteClusterPrivileges = translatedResponse.getRemoteClusterPrivileges();
+                    if (remoteClusterPrivileges.length > 0) { // remote clusters are not supported in stateless mode, so hide entirely
```

Comment:
Does this omit the field entirely if there aren't any available remote cluster privileges? From Kibana's (selfish) perspective, it'd be easier to handle an empty array than an optional field.

Reply:
as-is it will be completely hidden in serverless. Won't Kibana need special handling for serverless regardless (empty array or missing) ? e.g. don't present this at all in Kibana in serverless (at least not until we expose CCS) ? I don't have a strong preference (empty array or missing).

Reply:
Currently, in serverless, the request will fail if Kibana presents an empty `remote_cluster` array (i.e. `"remote_cluster":[]`) to Elasticsearch. We could potentially treat it the same way as we do the `run_as` field in serverless (simply accept and ignore it if it is empty - but fail if it's not).

Reply:
Not a strong position at all, just wanted to raise for consideration: if possible it would be great to instead omit `remote_cluster` from the response if it's not set. That way we don't have to have the `run_as` accept-empty-but-nothing-else workaround. @legrego would that work for Kibana or is an empty array still preferable?

Reply:
For this PR:

> as-is it will be completely hidden in serverless. Won't Kibana need special handling for serverless regardless (empty array or missing) ? e.g. don't present this at all in Kibana in serverless (at least not until we expose CCS) ? I don't have a strong preference (empty array or missing).

We have special handling for serverless to account for the different requirements. For this "Builtin privilege API", it is only used to populate the dropdown of available options. We hide the dropdown on serverless, so we won't look for the property.
For on-prem, we'll be looking for this property, and we'll need to add additional logic to test if it's present before using it. I'm really bike-shedding here, but in general it's easier from a type-safety perspective if we can rely on the fields being present in the response. As a recent example, our upgrade assistant was broken for ESS users because we expected a field to be present on role mappings which was absent: https://github.com/elastic/kibana/issues/177065


----

Now we're discussing a different set of Role APIs:

> Currently, in serverless, the request will fail if Kibana presents an empty remote_cluster array (i.e. "remote_cluster":[]) to Elasticsearch. We could potentially treat it the same way as we do the run_as field in serverless (simply accept and ignore it if it is empty - but fail if it's not).

@slobodanadamovic Is the different behavior between `remote_cluster` and `run_as` intentional?

> Not a strong position at all, just wanted to raise for consideration: if possible it would be great to instead omit remote_cluster from the response if it's not set. That way we don't have to have the run_as accept-empty-but-nothing-else workaround. @legrego would that work for Kibana or is an empty array still preferable?

(I assume we're talking about the PUT/POST role API) Is there a convention we should follow to distinguish between "not setting a value" vs "unsetting a value"?

One of our goals has been for Kibana to have as little knowledge of ES's role validation requirements as possible. This allows our systems to evolve more independently and reduces the number of times we need to release updates in lockstep. The knowledge that we do have is often gleaned from the type definitions that are automatically generated by the transport client. Having the same set of ES APIs manage two diverging sets of role definitions is at odds with this goal, as Kibana will have to understand when it is safe to send some properties, and not others.

This goal might not make sense anymore, so we can certainly revisit this if desired.

Reply:
> @slobodanadamovic Is the different behavior between remote_cluster and run_as intentional?

Apologies, I suggested it without knowing much about the context. The `run_as` field was there longer and I think the handling of it was intentional for BWC reason. After some discussions, seems like we should handle `remote_cluster` the same way we handle `remote_indices` field. In stateless we don't have any special handling for `remote_indices` and we treat it as an empty array. How is `remote_indices` handled on  Kibana side?

> (I assume we're talking about the PUT/POST role API)
Correct. The issue I mentioned is not related to the changes in this PR. We can decide how to handle and address it in followup if change on ES side is necessary.

>  Is there a convention we should follow to distinguish between "not setting a value" vs "unsetting a value"?

Unfortunately, I don't think there is one.